### ISSUE TITLE
Repo: Standardize debug namespaces per-file

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#5872]
### Suggestion

Splitting out of https://github.com/typescript-eslint/typescript-eslint/pull/5834#discussion_r1002821063:

> in an ideal world we'd actually do a unique substring per file to help differentiate where logs are coming from.
> 
> eg `typescript-eslint:typescript-estree:inferSingleRun`, `typescript-eslint:typescript-estree:create-program:useProvidedPrograms`, etc
> 
> this also allows you to easily filter the log lines to specific usecases with env vars like `DEBUG=typescript-eslint:typescript-estree:create-program:*`
> 
> also helpful because `debug` also colours each string differently IIRC so it's easy to see different sources when parsing logs.
> 
> I wonder if we can automate creation of the string via like
> 
> ```ts
> const PACKAGE_ROOT = "resolve from disk somehow";
> 
> function getLogger(filename: string) {
>   const relative = path.relative(PACKAGE_ROOT, filename);
>   const parsed = path.parse(relative);
>   return debug(`typescript-eslint:${parsed.dir.replace(/\//g, ':')}:${parsed.name}`);
> }
> 
> // then call this via
> const log = getLogger(__filename);
> ```

It'd be nifty if we could use a lint rule that enforces such a thing.

**Comment by bradzacher** [comment]
> It'd be nifty if we could use a lint rule that enforces such a thing.

Sadly [`eslint-plugin-debug`](https://www.npmjs.com/package/eslint-plugin-debug) doesn't exist :(

We could create an internal rule for us to enforce the style, for sure!

**Comment by JoshuaKGoldberg** [comment]
I tried out the `getLogger(__filename)` strategy in #10598. 

* Advantage: it's very convenient to no longer have to manually massage those strings!
* Disadvantage: the strings are no longer searchable

I think the lint rule will be the way to go. We can make provide auto-fixes so we don't have to manually type strings out.
